apis/Google.Cloud.Datastore.V1/Google.Cloud.Datastore.V1/DatastoreDbBuilder.cs

```diff
@@ -0,0 +1,247 @@
+                            GaxPreconditions.CheckState(emulatorConfig.ProjectId == null, "{0} should not be set when {1} is not", EmulatorConfiguration.EmulatorProjectVariable, EmulatorConfiguration.EmulatorHostVariable);
+                        }
+                        break;
+                }
+
+                void ApplyEmulatorConfiguration()
+                {
+                    // TODO: Do we want to require this? Or should a configured emulator override any credentials?
+                    GaxPreconditions.CheckState(clientBuilder.Endpoint == null, "{0} should not be set when connecting to an emulator", nameof(clientBuilder.Endpoint));
+                    GaxPreconditions.CheckState(clientBuilder.CallInvoker == null, "{0} should not be set when connecting to an emulator", nameof(clientBuilder.CallInvoker));
+                    GaxPreconditions.CheckState(clientBuilder.ChannelCredentials == null, "{0} should not be set when connecting to an emulator", nameof(clientBuilder.ChannelCredentials));
+                    GaxPreconditions.CheckState(clientBuilder.CredentialsPath == null, "{0} should not be set when connecting to an emulator", nameof(clientBuilder.CredentialsPath));
+                    GaxPreconditions.CheckState(clientBuilder.JsonCredentials == null, "{0} should not be set when connecting to an emulator", nameof(clientBuilder.JsonCredentials));
+                    GaxPreconditions.CheckState(clientBuilder.Scopes == null, "{0} should not be set when connecting to an emulator", nameof(clientBuilder.Scopes));
+                    GaxPreconditions.CheckState(clientBuilder.TokenAccessMethod == null, "{0} should not be set when connecting to an emulator", nameof(clientBuilder.TokenAccessMethod));
+                    clientBuilder.Endpoint = emulatorConfig.Endpoint;
+                    clientBuilder.ChannelCredentials = Grpc.Core.ChannelCredentials.Insecure;
+                    projectId = projectId ?? emulatorConfig.ProjectId;
+                }
+            }
+            GaxPreconditions.CheckState(!string.IsNullOrEmpty(projectId), "The project ID must be configured");
+            return new ConfiguredBuilder(projectId, NamespaceId, clientBuilder);
+        }
+
+        /// <summary>
+        /// The namespace ID, or null to use the default.
+        /// </summary>
+        public string NamespaceId { get; set; }
+
+        /// <inheritdoc />
+        public override DatastoreDb Build() =>
+            PrepareBuilder().Build();
+
+        /// <inheritdoc />
+        public override Task<DatastoreDb> BuildAsync(CancellationToken cancellationToken = default) =>
+            PrepareBuilder().BuildAsync(cancellationToken);
+
+        // We never end up using these methods, at least with the current implementation
+        /// <inheritdoc />
+        protected override ServiceEndpoint GetDefaultEndpoint() =>
+            throw new InvalidOperationException($"This method should never execute in DatastoreDbBuilder");
```

Comment:
nit: `$` not required (and following two methods)

Reply:
Thanks, will fix.